src/ModelContextProtocol/Client/SseClientSessionTransport.cs

```diff
@@ -92,26 +92,25 @@ public override async Task SendMessageAsync(
         StreamableHttpClientSessionTransport.CopyAdditionalHeaders(httpRequestMessage.Headers, _options.AdditionalHeaders);
         var response = await _httpClient.SendAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
 
-        response.EnsureSuccessStatusCode();
-
-        var responseContent = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
-
-        if (string.IsNullOrEmpty(responseContent) || responseContent.Equals("accepted", StringComparison.OrdinalIgnoreCase))
-        {
-            LogAcceptedPost(Name, messageId);
-        }
-        else
+        if (!response.IsSuccessStatusCode)
         {
             if (_logger.IsEnabled(LogLevel.Trace))
             {
-                LogRejectedPostSensitive(Name, messageId, responseContent);
+                LogRejectedPostSensitive(Name, messageId, await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false));
             }
             else
             {
                 LogRejectedPost(Name, messageId);
             }
 
-            throw new InvalidOperationException("Failed to send message");
+            response.EnsureSuccessStatusCode();
+        }
+
+        if (response.Content.Headers.ContentType?.MediaType is "application/json")
+        {
+            // Certain MCP servers implementing SSE may return the response in the current POST request instead of the SSE stream.
+            // Even though this is not officially part of the SSE protocol, we handle it here.
+            await ProcessInboundMessage(await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false), cancellationToken).ConfigureAwait(false);
```

Comment:
```suggestion
```

Did you try this with the [Haystack Search MCP Server](https://marketplace.visualstudio.com/items?itemName=codetrek.haystack-search)? In my testing, it sends the exact same message over the SSE channel leading to a log message about an "unknown request ID" since there's no longer an entry in the _pendingRequests ConcurrentDictionary for the initialize request.

```
info: ModelContextProtocol.Client.McpClient[429065462]
      http://localhost:13135/mcp/sse received request for unknown request ID '1'.
```

I think we should do the simpler thing and just ignore the response body. At most, I'd log the content at a trace level. And I think we should update the ..._ServerReturningResponseInPostRequest_... test to replicate the Haystack behavior.

Reply:
> In my testing, it sends the exact same message over the SSE channel leading to a log message about an "unknown request ID"

I take it you ran the test server as opposed to the Haystack server? The former was incorrectly issuing responses for notifications which is what was triggering those logs.

FWIW I just downloaded an update to Haystack after which I have stopped seeing JSON messages appear in responses. I presume this was a bug that got fixed on their side.

Unless @PederHP objects then, I'm going to completely remove the response content handling like you suggested.